% Generated by roxygen2: do not edit by hand
% Please edit documentation in R/srat_test.R
\name{lsrat_test}
\alias{lsrat_test}
\title{Calculate prospective and retrospective P-values for GEE or GLMM model}
\usage{
lsrat_test(
  srat.est,
  G,
  weights = "beta",
  impute.method = "mean",
  GRM = NULL,
  tests = c("B", "S", "O", "E", "V", "R"),
  B = 5000,
  rho = c(0, 0.5, 1),
  return_single = FALSE
)
}
\arguments{
\item{srat.est}{The output of function "lsrat_est()" or "rsmmat_est()"}

\item{G}{The genotype matrix of SNP-set, an m*q matrix where m is the number of subjects and q is the total number genetic variants.}

\item{impute.method}{choose the imputation method when there is missing genotype. Optional options are: 'random', 'fixed' or 'bestguess'.}

\item{GRM}{takes m-by-m genetic correlation matrix or kinship matrix.}

\item{tests}{a character vector indicating which LSRAT/RSMMAT tests should be performed ("B" for the burden test, "S" for SKAT, "V" for variant-level ACAT test, "O" for SKAT-O, "E" for the SMMAT test, "R" for omnibus ACAT test).}

\item{B}{a number of perturbation used for P-value approximation}

\item{rho}{a numeric vector defining the search grid used in SMMAT-O for SKAT-O (see the SKAT-O paper for details). Default  = c(0, 0.5, 1)}

\item{return_single}{Logical parameter indicating whether single variant P-value to be returned or not. Default = FALSE.}
}
\value{
This function returns a list of the specified tests, the prospective,  retrospective (and pertubed) P-values for the SNP-set of each tests is listed within each test.
}
\description{
This function tests a SNPs for a given SNP set for a given lbrat estimated null model.
}
